Mathlib/Tactic/LibraryRewrite.lean

```diff
@@ -0,0 +1,364 @@
+  forallTelescope cinfo.type fun _ e => do
+    let some (lhs, rhs) := matchEqn? e | throwError "Expected equation, not {indentExpr e}"
+    let side := if rwLemma.symm then rhs else lhs
+    ppExprTagged side
+
+structure RewriteApplication extends RewriteLemma where
+  tactic : String
+  replacement : CodeWithInfos
+  extraGoals : Array CodeWithInfos
+
+/-- Return `e` as a string for pasting into the editor. -/
+def toPasteString (e : Expr) : MetaM String :=
+  withOptions (·.setBool ``pp.universes false) do
+    return Format.pretty (← ppExpr e)
+
+def rewriteCall (loc : SubExpr.GoalsLocation) (rwLemma : RewriteLemma) :
+    MetaM (Option RewriteApplication) := do
+  let thm ← mkConstWithFreshMVarLevels rwLemma.name
+  let (mvars, bis, eqn) ← forallMetaTelescope (← inferType thm)
+  let some (lhs, rhs) := matchEqn? eqn | return none
+  let target ← loc.rootExpr
+  let subExpr ← Core.viewSubexpr loc.pos target
+  let (lhs, rhs) := if rwLemma.symm then (rhs, lhs) else (lhs, rhs)
+  unless ← isDefEq lhs subExpr do return none
+
+  let mut extraGoals := #[]
+  for mvar in mvars, bi in bis do
+    let mvarId := mvar.mvarId!
+    -- we need to check that all instances can be synthesized
+    if bi.isInstImplicit then
+      unless (← trySynthInstance (← mvarId.getType)) matches .some _ do
+        return none
+    else if bi.isExplicit then
+      if !(← mvarId.isAssigned) then
+        -- if the userName has macro scopes, we can't use the name, so we use `?_` instead
+        if (← mvarId.getDecl).userName.hasMacroScopes then
+          mvarId.setUserName `«_»
+        let extraGoal ← instantiateMVars (← mvarId.getType)
+        extraGoals := extraGoals.push (← ppExprTagged extraGoal)
+
+  let replacement ← ppExprTagged (← instantiateMVars rhs)
+  let lemmaApplication ← instantiateMVars (mkAppN thm mvars)
+
+  let location ← (do match loc.loc with
+    | .hyp fvarId
+    | .hypType fvarId _ => return s! " at {← fvarId.getUserName}"
+    | _ => return "")
+  let symm := if rwLemma.symm then "← " else ""
+  let lhs ← instantiateMVars lhs
+  let positions ← findPositions lhs target
+  let cfg := match positions.findIdx? (· == loc.pos) with
+    | none => " /- Error: couldn't find a suitable occurrence -/"
+    | some pos =>
+      if positions.size == 1 then "" else
+        s! " (config := \{ occs := .pos [{pos+1}]})"
+  let tactic := s! "rw{cfg} [{symm}{← toPasteString lemmaApplication}]{location}"
+  return some { rwLemma with tactic, extraGoals, replacement }
+
+
+def renderResults (results : Array (CodeWithInfos × Array RewriteApplication)) (isEverything : Bool)
+    (range : Lsp.Range) (doc : FileWorker.EditableDocument) : Html :=
+  let htmls := results.map (fun (t, arr) => renderBlock t arr)
+  let title := s! "{if isEverything then "All" else "Some"} rewrite suggestions:";
+  <details «open»={true}>
+    <summary className="mv2 pointer"> {.text title}</summary>
+    {.element "div" #[("style", json% {"margin-left" : "4px"})] htmls}
```

Comment:
```suggestion
    {.element "div" #[("style", json% {"marginLeft" : "4px"})] htmls}
```
We pass the props to React, so React-style names should be used.

Reply:
Ok, I've made the changes